File: app/src/main/java/ru/nikshlykov/englishwordsapp/di/DatabaseModule.java
package ru.nikshlykov.englishwordsapp.di;

import android.app.Application;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import ru.nikshlykov.englishwordsapp.db.GroupsRepository;
import ru.nikshlykov.englishwordsapp.db.ModesRepository;
import ru.nikshlykov.englishwordsapp.db.WordsRepository;

@Module
public class DatabaseModule {

    @Provides
    @Singleton
    ModesRepository provideModesRepository(Application application){
        return new ModesRepository(application);
    }

    @Provides
    @Singleton
    GroupsRepository provideGroupsRepository(Application application){
        return new GroupsRepository(application);
    }

    @Provides
    @Singleton
    WordsRepository provideWordsRepository(Application application){
        return new WordsRepository(application);
    }
}

File: app/src/main/java/ru/nikshlykov/englishwordsapp/di/DatabaseModule.kt
package ru.nikshlykov.englishwordsapp.di

import android.content.Context
import dagger.Module
import dagger.Provides
import ru.nikshlykov.data.database.DaoProvider
import ru.nikshlykov.data.database.daos.GroupDao
import ru.nikshlykov.data.database.daos.LinkDao
import ru.nikshlykov.data.database.daos.ModeDao
import ru.nikshlykov.data.database.daos.RepeatDao
import ru.nikshlykov.data.database.daos.SubgroupDao
import ru.nikshlykov.data.database.daos.WordDao
import javax.inject.Singleton

@Module
class DatabaseModule {
    @Provides
    @Singleton
    fun provideDaoProvider(context: Context): DaoProvider {
        return DaoProvider(context)
    }

    @Provides
    @Singleton
    fun provideSubgroupDao(daoProvider: DaoProvider): SubgroupDao {
        return daoProvider.getSubgroupDao()
    }

    @Provides
    @Singleton
    fun provideGroupDao(daoProvider: DaoProvider): GroupDao {
        return daoProvider.getGroupDao()
    }

    @Provides
    @Singleton
    fun provideModeDao(daoProvider: DaoProvider): ModeDao {
        return daoProvider.getModeDao()
    }

    @Provides
    @Singleton
    fun provideLinkDao(daoProvider: DaoProvider): LinkDao {
        return daoProvider.getLinkDao()
    }

    @Provides
    @Singleton
    fun provideWordDao(daoProvider: DaoProvider): WordDao {
        return daoProvider.getWordDao()
    }

    @Provides
    @Singleton
    fun provideRepeatDao(daoProvider: DaoProvider): RepeatDao {
        return daoProvider.getRepeatDao()
    }
}